spyder_kernels/utils/iofuncs.py

```diff
@@ -288,58 +289,101 @@ def save_dictionary(data, filename):
+                        saved_arrays[(name, None)] = osp.basename(fname)
+                        data.pop(name)
+                    elif isinstance(data[name], (list, dict)):
+                        # Save arrays nested in lists or dictionaries
+                        if isinstance(data[name], list):
+                            iterator = enumerate(data[name])
+                        else:
+                            iterator = iter(list(data[name].items()))
+                        to_remove = []
+                        for index, value in iterator:
+                            if isinstance(value,
+                                          np.ndarray) and value.size > 0:
+                                fname = __save_array(value, arr_fname,
+                                                     len(saved_arrays))
+                                saved_arrays[(name, index)] = (
+                                    osp.basename(fname))
+                                to_remove.append(index)
+                        for index in sorted(to_remove, reverse=True):
+                            data[name].pop(index)
+                    # If an array can't be saved with numpy for some reason,
+                    # leave the object intact and try to save it normally.
+                except (RuntimeError, pickle.PicklingError, TypeError):
+                    pass
             if saved_arrays:
                 data['__saved_arrays__'] = saved_arrays
-        pickle_filename = osp.splitext(filename)[0]+'.pickle'
+
+        pickle_filename = osp.splitext(filename)[0] + '.pickle'
+        # Attempt to pickle everything.
+        # If pickling fails, iterate through to eliminate problem objs & retry.
         with open(pickle_filename, 'w+b') as fdesc:
-            pickle.dump(data, fdesc, 2)
-        tar = tarfile.open(filename, "w")
-        for fname in [pickle_filename]+[fn for fn in list(saved_arrays.values())]:
-            tar.add(osp.basename(fname))
-            os.remove(fname)
-        tar.close()
-        if saved_arrays:
-            data.pop('__saved_arrays__')
+            try:
+                pickle.dump(data, fdesc, protocol=2)
+            except (pickle.PicklingError, AttributeError, TypeError,
+                    ImportError, IndexError, RuntimeError):
+                data_filtered = {}
+                for obj_name, obj_value in data.items():
+                    try:
+                        pickle.dumps(obj_value, protocol=2)
+                    except Exception:
+                        skipped_keys.append(obj_name)
+                    else:
+                        data_filtered[obj_name] = obj_value
+                if not data_filtered:
+                    raise RuntimeError('No supported objects to save')
+                pickle.dump(data_filtered, fdesc, protocol=2)
+
+        # Use PAX (POSIX.1-2001) format instead of default GNU.
+        # This improves interoperability and UTF-8/long variable name support.
```

Comment:
Is this change backwards compatible? I mean, what will happen with old spydata files? Will users be able to open them after this change?

Reply:
Yup, 100%—I made sure of that. It has no effect on reading since any remotely modern version of the standard library ``tarfile`` module dating back to at least before Python 2.6 (which Spyder no longer supports anyway) can read PAX as well as the default GNU format, and the proper format to read it as detected automatically from the file when loading. It just means any new files saved will be a little less restrictive and more interoperable.

Reply:
Ok, good to know.